external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -18,228 +18,253 @@
    */
-  def get(
-      offset: Long,
-      untilOffset: Long,
-      pollTimeoutMs: Long,
-      failOnDataLoss: Boolean): ConsumerRecord[Array[Byte], Array[Byte]] = {
-    internalConsumer.get(offset, untilOffset, pollTimeoutMs, failOnDataLoss)
+  def fetch(offset: Long, pollTimeoutMs: Long)
+  : (ju.List[ConsumerRecord[Array[Byte], Array[Byte]]], Long) = {
+    // Seek to the offset because we may call seekToBeginning or seekToEnd before this.
+    seek(offset)
+    val p = consumer.poll(pollTimeoutMs)
+    val r = p.records(topicPartition)
+    logDebug(s"Polled $groupId ${p.partitions()}  ${r.size}")
+    val offsetAfterPoll = consumer.position(topicPartition)
+    logDebug(s"Offset changed from $offset to $offsetAfterPoll after polling")
+    val fetchedData = (r, offsetAfterPoll)
+    if (r.isEmpty) {
+      // We cannot fetch anything after `poll`. Two possible cases:
+      // - `offset` is out of range so that Kafka returns nothing. `OffsetOutOfRangeException` will
+      //   be thrown.
+      // - Cannot fetch any data before timeout. `TimeoutException` will be thrown.
+      // - Fetched something but all of them are not invisible. This is a valid case and let the
+      //   caller handles this.
+      val range = getAvailableOffsetRange()
+      if (offset < range.earliest || offset >= range.latest) {
+        throw new OffsetOutOfRangeException(
+          Map(topicPartition -> java.lang.Long.valueOf(offset)).asJava)
+      } else if (offset == offsetAfterPoll) {
+        throw new TimeoutException(
+          s"Cannot fetch record for offset $offset in $pollTimeoutMs milliseconds")
+      }
+    }
+    fetchedData
   }
 
   /**
    * Return the available offset range of the current partition. It's a pair of the earliest offset
    * and the latest offset.
    */
-  def getAvailableOffsetRange(): AvailableOffsetRange = internalConsumer.getAvailableOffsetRange()
+  def getAvailableOffsetRange(): AvailableOffsetRange = {
+    consumer.seekToBeginning(Set(topicPartition).asJava)
+    val earliestOffset = consumer.position(topicPartition)
+    consumer.seekToEnd(Set(topicPartition).asJava)
+    val latestOffset = consumer.position(topicPartition)
+    AvailableOffsetRange(earliestOffset, latestOffset)
+  }
 
-  /**
-   * Release this consumer from being further used. Depending on its implementation,
-   * this consumer will be either finalized, or reset for reuse later.
-   */
-  def release(): Unit
+  override def close(): Unit = {
+    consumer.close()
+  }
 
-  /** Reference to the internal implementation that this wrapper delegates to */
-  protected def internalConsumer: InternalKafkaConsumer
-}
+  /** Create a KafkaConsumer to fetch records for `topicPartition` */
+  private def createConsumer: KafkaConsumer[Array[Byte], Array[Byte]] = {
+    val updatedKafkaParams = KafkaConfigUpdater("executor", kafkaParams.asScala.toMap)
+      .setAuthenticationConfigIfNeeded()
```

Comment:
Not sure whether we have to change or not but would like to start a discussion.
Basically there are 2 ways to store consumers with delegation token:
* Token is part of the cache key
* Token is not part of the cache key

I've chosen the latter in the producer area because there the eviction is time based so old/not used consumers are evicted by default every 10 minutes. The main gain would be not to wait a consumer failure + task restart to apply the new token.

Not yet sure worth to do the same here because I have to take a look at the commons pool eviction policy again...

Reply:
I would need to have some time to revisit my patch to refresh context to consider this. The patch was sitting around an year so even feels new to me. :)

Reply:
Same situation, just wanted to highlight that we need to consider it :)

Reply:
If Kafka provides some kind of "ping-pong" mechanism we can leverage it in multiple places, verifying idle objects, verifying objects before borrowing, etc.

`PoolObjectFactory` has `validateObject` for this. Please refer below line to see example: 

https://github.com/xetorthio/jedis/blob/9504d7141059dbcd89d28ae2d0f5b76932d754f6/src/main/java/redis/clients/jedis/JedisFactory.java#L141-L156



Not sure it is available in Kafka client.

Reply:
I've made a deep dive and thought the situation through and here are my findings.
* The original behavior is not changed, which I think it's fine here. Please see my points below.
* Putting the token into the cache key is not a good idea because if the eviction is not time based then the cache will constantly grow till it's max size and stay it like that. This would introduce quite a constraint and in case of large cluster it could be quite a waste of resources. => As a conclusion I'm fine with the actual way and later this can be enhanced in a different jira.
* I've had a look at the "ping-pong" proposal and I think it could generate quite a traffic. I think this can be solved in a more cheap way.
* What I think would be a better solution is to invalidate (but not close if used) consumers when the delegation token changed. This way tasks can pick up the latest and greatest token + not needed to do and additional administration.

All in all this discussion can be resolved from my perspective.

Reply:
> Do you have any idea to check whether token for current consumer is invalid and new token is available?

To answer this question I don't know any way to detect such situation. The only way to detect it is the thrown exception. If any kind of exception happens then the consumer has to be invalidated (maybe the exception is not even token related).

Reply:
OK so if I understand correctly, we expect task failure for this case, and we will invalidate consumers for the key so it would work, though it may not be ideal. (Not sure we can do reacquire in this case, but without knowing the exception of delegation token we can't simply reacquire and retry.)
I'll mark this as resolved. Please reopen if I'm missing here. Thanks for the analyze!